controlplane/kubeadm/internal/etcd/etcd.go

```diff
@@ -0,0 +1,233 @@
+	AlarmCorrupt
+)
+
+// Adapted from kubeadm
+
+// Member struct defines an etcd member; it is used to avoid spreading
+// github.com/coreos/etcd dependencies.
+type Member struct {
+	// ID is the ID of this cluster member
+	ID uint64
+
+	// Name is the human-readable name of the member. If the member is not started, the name will be an empty string.
+	Name string
+
+	// PeerURLs is the list of URLs the member exposes to the cluster for communication.
+	PeerURLs []string
+
+	// ClientURLs is the list of URLs the member exposes to clients for communication. If the member is not started, clientURLs will be empty.
+	ClientURLs []string
+
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+
+	// Alarms is the list of alarms for a member.
+	Alarms []AlarmType
+}
+
+// pbMemberToMember converts the protobuf representation of a cluster member to a Member struct.
+func pbMemberToMember(m *etcdserverpb.Member) *Member {
+	return &Member{
+		ID:         m.GetID(),
+		Name:       m.GetName(),
+		PeerURLs:   m.GetPeerURLs(),
+		ClientURLs: m.GetClientURLs(),
+		IsLearner:  m.GetIsLearner(),
+		Alarms:     []AlarmType{},
+	}
+}
+
+// NewEtcdClient creates a new etcd client with a custom dialer and is configuration with optional functions.
+func NewEtcdClient(endpoint string, dialer GRPCDial, tlsConfig *tls.Config) (*clientv3.Client, error) {
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Endpoints:   []string{endpoint},
+		DialTimeout: etcdTimeout,
+		DialOptions: []grpc.DialOption{
+			grpc.WithBlock(), // block until the underlying connection is up
+			grpc.WithContextDialer(dialer),
+		},
+		TLS: tlsConfig,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create etcd client")
+	}
+	etcdClient.Endpoints()
+	return etcdClient, nil
+}
+
+// NewClientWithEtcd configures our response formatter (Client) with an etcd client and endpoint.
+func NewClientWithEtcd(etcdClient etcd) (*Client, error) {
+	if len(etcdClient.Endpoints()) == 0 {
+		return nil, errors.New("etcd client was not configured with any endpoints")
+	}
+	return &Client{
+		Endpoint:   etcdClient.Endpoints()[0],
+		EtcdClient: etcdClient,
+	}, nil
+}
+
+// Close closes the etcd client.
+func (c *Client) Close() error {
+	return c.EtcdClient.Close()
+}
+
+// Member retrieves the Member information for a given peer.
+func (c *Client) Member(ctx context.Context, peerURL string) (*Member, error) {
+	members, err := c.Members(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, m := range members {
+		if m.PeerURLs[0] == peerURL {
+			return m, nil
+		}
+	}
+	return nil, errors.Errorf("etcd member %v not found", peerURL)
+}
+
+// Members retrieves a list of etcd members.
+func (c *Client) Members(ctx context.Context) ([]*Member, error) {
+	response, err := c.EtcdClient.MemberList(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get list of members for etcd cluster")
+	}
+
+	alarms, err := c.Alarms(ctx)
```

Comment:
The downside is that this makes a separate API call. However, it is one call, not N calls (where N is the number of members).

The upside is that the user does not need to query Alarms separately and correlate entries returned by `Alarms` to entries in the list of `Members` by using the MemberID.

I don't have strong feelings either way.

Reply:
Yep, I am thinking of *Client in this package as basically a thing that uses the etcd client to fetch data to populate some structure that we want. It's the place I would expect to wrap multiple API calls to build out a data structure like this so I think this call belongs here. Are you worried about an additional API call for performance reasons?